stf/src/stf_sgx_primitives.rs

```diff
@@ -0,0 +1,152 @@
+		"System",
+		"Account",
+		who,
+		&StorageHasher::Blake2_128Concat,
+	)) {
+		if let Ok(info) = AccountInfo::decode(&mut infovec.as_slice()) {
+			Some(info)
+		} else {
+			None
+		}
+	} else {
+		None
+	}
+}
+
+pub fn validate_nonce(who: &AccountId, nonce: Index) -> StfResult<()> {
+	// validate
+	let expected_nonce = get_account_info(who).map_or_else(|| 0, |acc| acc.nonce);
+	if expected_nonce == nonce {
+		return Ok(())
+	}
+	Err(StfError::InvalidNonce(nonce))
+}
+
+/// increment nonce after a successful call execution
+pub fn increment_nonce(account: &AccountId) {
+	//FIXME: Proper error handling - should be taken into
+	// consideration after implementing pay fee check
+	if let Some(mut acc_info) = get_account_info(account) {
+		debug!("incrementing account nonce");
+		acc_info.nonce += 1;
+		sp_io::storage::set(&account_key_hash(account), &acc_info.encode());
+		debug!(
+			"updated account {:?} nonce: {:?}",
+			account.encode(),
+			get_account_info(account).unwrap().nonce
+		);
+	} else {
+		error!("tried to increment nonce of a non-existent account")
+	}
+}
+
+pub fn storage_value_key(module_prefix: &str, storage_prefix: &str) -> Vec<u8> {
+	let mut bytes = sp_core::twox_128(module_prefix.as_bytes()).to_vec();
+	bytes.extend(&sp_core::twox_128(storage_prefix.as_bytes())[..]);
+	bytes
+}
+
+pub fn storage_map_key<K: Encode>(
+	module_prefix: &str,
+	storage_prefix: &str,
+	mapkey1: &K,
+	hasher1: &StorageHasher,
+) -> Vec<u8> {
+	let mut bytes = sp_core::twox_128(module_prefix.as_bytes()).to_vec();
+	bytes.extend(&sp_core::twox_128(storage_prefix.as_bytes())[..]);
+	bytes.extend(key_hash(mapkey1, hasher1));
+	bytes
+}
+
+pub fn storage_double_map_key<K: Encode, Q: Encode>(
+	module_prefix: &str,
+	storage_prefix: &str,
+	mapkey1: &K,
+	hasher1: &StorageHasher,
+	mapkey2: &Q,
+	hasher2: &StorageHasher,
+) -> Vec<u8> {
+	let mut bytes = sp_core::twox_128(module_prefix.as_bytes()).to_vec();
+	bytes.extend(&sp_core::twox_128(storage_prefix.as_bytes())[..]);
+	bytes.extend(key_hash(mapkey1, hasher1));
+	bytes.extend(key_hash(mapkey2, hasher2));
+	bytes
+}
+
+/// generates the key's hash depending on the StorageHasher selected
+fn key_hash<K: Encode>(key: &K, hasher: &StorageHasher) -> Vec<u8> {
+	let encoded_key = key.encode();
+	match hasher {
+		StorageHasher::Identity => encoded_key.to_vec(),
+		StorageHasher::Blake2_128 => sp_core::blake2_128(&encoded_key).to_vec(),
+		StorageHasher::Blake2_128Concat => {
+			// copied from substrate Blake2_128Concat::hash since StorageHasher is not public
+			let x: &[u8] = encoded_key.as_slice();
+			sp_core::blake2_128(x).iter().chain(x.iter()).cloned().collect::<Vec<_>>()
+		},
+		StorageHasher::Blake2_256 => sp_core::blake2_256(&encoded_key).to_vec(),
+		StorageHasher::Twox128 => sp_core::twox_128(&encoded_key).to_vec(),
+		StorageHasher::Twox256 => sp_core::twox_256(&encoded_key).to_vec(),
+		StorageHasher::Twox64Concat => sp_core::twox_64(&encoded_key).to_vec(),
+	}
+}
```

Comment:
Nice, this has been bugging me for ages in the original sgx.rs file.

Reply:
I had to extract it, because the `test_genesis.rs` depends on it - so it couldn't remain in `sgx.rs`, otherwise we'd have a cyclic dependency (since `init_state()` in `sgx.rs` uses `test_genesis.rs`)

Reply:
Ah I see. I want to make these an entirely separate crate anyhow. I even believe I did this in my working branch.

Because, the api-client implements the same functions.

Reply:
I only talk about the hash functions.

Reply:
Can you create an issue for that? I think this is outside the scope of this PR. There are certainly plenty of improvements we can/should do - but this PR is about the 'test' feature in STF and the enclave.

Reply:
Yes, absolutely. It is not my intention to solve that in this PR. It needs some more thinking anyhow, because this is a cross-repo issue. I amended this issue to incorporate the storage-hasher aspect: https://github.com/integritee-network/worker/issues/323